packages/compose/src/hooks/use-media-query/index.js

```diff
@@ -10,28 +28,30 @@ import { useState, useEffect } from '@wordpress/element';
  * @return {boolean} return value of the media query.
  */
 export default function useMediaQuery( query ) {
-	const [ match, setMatch ] = useState(
-		() =>
-			!! (
-				query &&
-				typeof window !== 'undefined' &&
-				window.matchMedia( query ).matches
-			)
+	const subscribe = useCallback(
+		/** @type {(onStoreChange: () => void) => () => void} */
+		( onStoreChange ) => {
+			const mediaQueryList = getMediaQueryList( query );
+			if ( ! mediaQueryList ) {
+				return () => {};
+			}
+
+			mediaQueryList.addEventListener( 'change', onStoreChange );
+			return () => {
+				mediaQueryList.removeEventListener( 'change', onStoreChange );
+			};
+		},
+		[ query ]
 	);
 
-	useEffect( () => {
-		if ( ! query ) {
-			return;
-		}
-		const updateMatch = () =>
-			setMatch( window.matchMedia( query ).matches );
-		updateMatch();
-		const list = window.matchMedia( query );
-		list.addListener( updateMatch );
-		return () => {
-			list.removeListener( updateMatch );
-		};
-	}, [ query ] );
+	const matches = useSyncExternalStore(
+		subscribe,
+		() => {
+			const mediaQueryList = getMediaQueryList( query );
```

Comment:
It's bad that each call of the `getValue` callback creates a new `MediaQueryList` from scratch, and that the object we're reading the `.matches` value from and the object to which we are subscribing are different. What about:
```js
function useMediaQuery( query ) {
  const source = useMemo( () => {
    const mediaQueryList = window.matchMedia( query );
    return {
      subscribe( listener ) {
        mediaQueryList.addEventListener( 'change', listener );
        return () => mediaQueryList.removeEventListener( 'change', listener );
      },
      getValue() {
        return mediaQueryList.matches;
      },
    };
  }, [ query ] );

  return useSyncExternalStore( source.subscribe, source.getValue, () => false );
}
```
This hook, every time `query` changes, creates a new `mediaQueryList` object and a "source" wrapper around it.

The only thing that's missing is handling `null` `query` and missing support for `matchMedia`.

Reply:
Thank you, @jsnajdr!

I had a suspicion that calling `getMediaQueryList` again in the snapshot callback wasn't ideal. The good thing is that I already have the `useMemo` version, and I will push that in a second.